### PR TITLE
Flexible component deployment

### PR DESCRIPTION
**Pull request by shylesh**
Make components like rgw, cephfs, noobaa, cephblockpool
to be flexibly deplyed.

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

### CODE REVIEW

**Comment by shylesh** [comment]
@clacroix12 @vavuthu @dahorak @vasukulkarni ,
With this PR we are introducing a new cli option `--disable-components` followed by component names to be disabled from deployment. Currently this PR implements a comma separated component names  like 
```python
--disable-components rgw,cephfs,blockpools
```

However @petr-balogh suggested we go with 
```python
--disable-component rgw --disable-components cephfs --disable-component blockpool
```

which one you guys would feel more intuitive and better approach

**Comment by dahorak** [comment]
The first (currently implemented) option is shorter and I would vote for it. But if it will be (easily) possible to use both ways together (e.g. `--disable-components rgw,cephfs --disable-component blockpool`), it might be handy in some cases.

File: ocs_ci/framework/main.py
Comment (dahorak) [inline ocs_ci/framework/main.py:71]: If I remember it correctly, the `nargs="+"` means, that you can use multiple values for one option (space separated) - e.g.:
```
--disable-components rgw cephfs blockpool
```
I would not recommend this type of handling as it might be slightly unclear in our case (what is the parameter value and what is next parameter).

Comment (clacroix12) [reply ocs_ci/framework/main.py:71]: > @clacroix12 @vavuthu @dahorak @vasukulkarni ,
> With this PR we are introducing a new cli option `--disable-components` followed by component names to be disabled from deployment. Currently this PR implements a comma separated component names like
> 
> ```python
> --disable-components rgw,cephfs,blockpools
> ```
> 
> However @petr-balogh suggested we go with
> 
> ```python
> --disable-component rgw --disable-components cephfs --disable-component blockpool
> ```
> 
> which one you guys would feel more intuitive and better approach

We have other CLI options in which we use the append action (what petr suggested). I think we should be consistent and not introduce different means of adding multiple values to CLI options. One [example](https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/framework/main.py#L61) of this is `--ocsci-conf`. Because we configure it to use the append action, we can call `--ocsci-conf` multiple times and the values get added to a list automatically.

**Comment by clacroix12** [comment]
@shylesh just for consistency i would rename the config files to `disable_<component>.yaml` instead of `skip_<component>.yaml` Since enable/disable is the terminology we are using here, we should use the same in the filenames.

**Comment by shylesh** [comment]
rebased on latest master

File: conf/ocsci/skip_blockpools.yaml
Comment (vavuthu) [inline conf/ocsci/skip_blockpools.yaml:2]: If you are changing the filename to disable_blockpools.yaml, then it would be good have parameter disable_blockpools: True

File: ocs_ci/framework/main.py
Comment (dahorak) [reply ocs_ci/framework/main.py:71]: The consistency with `--ocsci-conf` is good point and I fully support this option - that's really good point. That means, that you can repeat the whole combination of parameter name and value - e.g. `--ocsci-conf config1.yaml --ocsci-conf config2.yaml`, which is enabled via `action="append"` parameter in `parser.add_argument(...)` ([python doc](https://docs.python.org/3/library/argparse.html#action), [usage in ocs-ci](https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/framework/main.py#L61)).
But my point in the comment above was against using `nargs="+"`, which will allow you to put multiple values to single parameter like this: `--ocsci-conf config1.yaml config2.yaml` ([python doc](https://docs.python.org/3/library/argparse.html#nargs). This syntax is might be slightly confusing to distinguish what is positional parameter and what is value for named parameter, so if it is not necessary, I would avoid it.

**Comment by dahorak** [comment]
> The first (currently implemented) option is shorter and I would vote for it. But if it will be (easily) possible to use both ways together (e.g. `--disable-components rgw,cephfs --disable-component blockpool`), it might be handy in some cases.

Coady have good point to keep it consistent with other parameters used in `ocs-ci`, so I agree with having only the second option (repeat whole combination of named argument and it's value) as it is for `--ocsci-conf` parameter.

File: ocs_ci/framework/main.py
Comment (clacroix12) [reply ocs_ci/framework/main.py:71]: @dahorak I totally agree with you. I must have replied to your comment instead of making a new one myself.

File: ocs_ci/templates/ocs-deployment/storage-cluster.yaml
Comment (petr-balogh) [inline ocs_ci/templates/ocs-deployment/storage-cluster.yaml:8]: This is just from 4.8 and this file is used also in other version.

Please check if we deploy via UI if those are even defined, I think we are trying to keep this file as minimal as possible and this should use the default from rook operator if not defined in storage cluster CR

File: ocs_ci/deployment/deployment.py
Comment (clacroix12) [inline ocs_ci/deployment/deployment.py:489]: I assume this is leftover from local testing?

File: ocs_ci/framework/__init__.py
Comment (clacroix12) [inline ocs_ci/framework/__init__.py:31]: We should document this new config section [here](https://github.com/red-hat-storage/ocs-ci/blob/master/conf/README.md)

File: ocs_ci/deployment/deployment.py
Comment (shylesh) [reply ocs_ci/deployment/deployment.py:489]: oops, yes , I missed it.

File: ocs_ci/framework/__init__.py
Comment (shylesh) [reply ocs_ci/framework/__init__.py:31]: ack, I will do that

File: ocs_ci/ocs/constants.py
Comment (petr-balogh) [inline ocs_ci/ocs/constants.py:1289]: ah, so capital letter was the problem? :)

File: ocs_ci/framework/__init__.py
Comment (petr-balogh) [reply ocs_ci/framework/__init__.py:31]: still missing documentation

**Review by petr-balogh — COMMENTED** [review]
Looks OK, after adding doc I will approve

File: ocs_ci/framework/__init__.py
Comment (shylesh) [reply ocs_ci/framework/__init__.py:31]: done

**Review by petr-balogh — APPROVED** [review]

File: ocs_ci/framework/main.py
Comment (dahorak) [inline ocs_ci/framework/main.py:74]: Since it takes one _component_ peer argument, I would use singular in the argument name: `--disable-component`.
Also it will be good to mention in the help, that this argument can be used multiple times.

**Review by dahorak — APPROVED** [review]
Just small suggestion about the argument name and help, otherwise LGTM.

**Review by clacroix12 — DISMISSED** [review]
LGTM! I agree with @dahorak about making the argument singular and adding to the help text but that can probably be done in a quick follow up PR. Nice work!

**Comment by shylesh** [comment]
rebased on latest master

**Review by petr-balogh — DISMISSED** [review]

**Review by clacroix12 — DISMISSED** [review]

**Review by petr-balogh — DISMISSED** [review]
Just thinking if we even need config files when we have cli argument? Maybe we can delete them and keep it cli only? Wdyt?

**Comment by shylesh** [comment]
> Just thinking if we even need config files when we have cli argument? Maybe we can delete them and keep it cli only? Wdyt?

@petr-balogh yes we can do that but I was just thinking about jenkins jobs. Do we need check boxes for components ? for now we can use these files to skip components for testing and once we have jenkins jobs with selectors we can remove this file. Even now another way is free text box for the job where we can pass this config. Let me know what's the best approach.

**Comment by petr-balogh** [comment]
> > Just thinking if we even need config files when we have cli argument? Maybe we can delete them and keep it cli only? Wdyt?
> 
> 
> 
> @petr-balogh yes we can do that but I was just thinking about jenkins jobs. Do we need check boxes for components ? for now we can use these files to skip components for testing and once we have jenkins jobs with selectors we can remove this file. Even now another way is free text box for the job where we can pass this config. Let me know what's the best approach.

We don't have jobs yet, so make jobs with new check boxes and do it this way directly. We have option to add additional pytest params for testing if we need it now

**Comment by shylesh** [comment]
rebased on the latest master

**Review by petr-balogh — APPROVED** [review]

**Comment by shylesh** [comment]
Verified locally with vsphere.

```html
16:28:04 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage get CephFilesystem  -n openshift-storage -o yaml
16:28:04 - MainThread - root - WARNING - list index out of range
16:28:04 - MainThread - root - WARNING - No CephFS found
16:28:04 - MainThread - ocs_ci.helpers.sanity_helpers - INFO - Deleting resources as a sanity functional validation
PASSED                                                                                                                           [100%]
---------------------------------------------------------- live log teardown -----------------------------------------------------------
16:28:05 - MainThread - tests.conftest - WARNING - During test were raised new alerts
16:28:05 - MainThread - tests.conftest - WARNING - [{'labels': {'alertname': 'KubeAPIErrorBudgetBurn', 'long': '3d', 'severity': 'warning', 'short': '6h'}, 'annotations': {'description': 'The API server is burning too much error budget.', 'summary': 'The API server is burning too much error budget.'}, 'state': 'pending', 'activeAt': '2021-05-18T23:28:05.333788348Z', 'value': '2.319061812430001e-02'}, {'labels': {'alertname': 'KubePodCrashLooping', 'container': 'prometheus', 'endpoint': 'https-main', 'job': 'kube-state-metrics', 'namespace': 'openshift-monitoring', 'pod': 'prometheus-k8s-0', 'service': 'kube-state-metrics', 'severity': 'warning'}, 'annotations': {'description': 'Pod openshift-monitoring/prometheus-k8s-0 (prometheus) is restarting 0.52 times / 10 minutes.', 'summary': 'Pod is crash looping.'}, 'state': 'pending', 'activeAt': '2021-05-18T23:25:46.474542648Z', 'value': '5.245166666666666e-01'}]
16:28:05 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc wait --for condition=ready pod -l app=rook-ceph-tools -n openshift-storage --timeout=120s
16:28:06 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage get pod -l 'app=rook-ceph-tools' -o jsonpath='{.items[0].metadata.name}'
16:28:06 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage exec rook-ceph-tools-5cfc5bbd87-h7cld -- ceph health
16:28:08 - MainThread - ocs_ci.utility.utils - INFO - Ceph cluster health is HEALTH_OK.
16:28:08 - MainThread - tests.conftest - INFO - Ceph health check passed at teardown
```

Due to LSO issue I couldn't verify this with LSO.

**Review by clacroix12 — APPROVED** [review]
LGTM!

**Comment by shylesh** [comment]
rebased on latest master

**Review by petr-balogh — DISMISSED** [review]

**Review by clacroix12 — APPROVED** [review]

**Comment by petr-balogh** [comment]
@shylesh I see there is again some conflict :( so please rebase again. Once you will get approval please merge, so we will have it in. Thanks

**Comment by shylesh** [comment]
rebased on latest master

**Review by clacroix12 — APPROVED** [review]

**Review by vasukulkarni — APPROVED** [review]

**Pull request merged**
